src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        decoder_attention_mask (:obj:`torch.BoolTensor` of shape :obj:`(batch_size, tgt_seq_len)`, `optional`, defaults to :obj:`None`):
+            Default behavior: generate a tensor that ignores pad tokens in decoder_input_ids. Causal mask will also be used by default.
+            If you want to change padding behavior, you should read :func:`~transformers.modeling_bart._prepare_decoder_inputs` and modify.
+            See diagram 1 in the paper for more info on the default strategy
+        output_attentions (:obj:`bool`, `optional`, defaults to :obj:`None`):
+            If set to ``True``, the attentions tensors of all attention layers are returned. See ``attentions`` under returned tensors for more detail.
+"""
+
+class ProphetNetPreTrainedModel(PreTrainedModel):
+    config_class = ProphetNetConfig
+    base_model_prefix = "model"
+
+    def _init_weights(self, module):
+        std = self.config.init_std
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+
+class LearnedPositionalEmbedding(nn.Embedding):
+    """
+    This module learns positional embeddings up to a fixed maximum size.
+    Padding ids are ignored by either offsetting based on padding_idx
+    or by setting padding_idx to None and ensuring that the appropriate
+    position ids are passed to the forward function.
+    """
+
+    def __init__(
+            self,
+            num_embeddings: int,
+            embedding_dim: int,
+            padding_idx: int,
+    ):
+        super().__init__(num_embeddings, embedding_dim, padding_idx)
+        self.onnx_trace = False
+
+    def forward(self, input, use_cache=False, positions=None):
+        """Input is expected to be of size [bsz x seqlen]."""
+        assert (
+            (positions is None) or (self.padding_idx is None)
+        ), "If positions is pre-computed then padding_idx should not be set."
+
+        if positions is None:
+            if use_cache:
+                # positions is the same for every token when decoding a single step
+                # Without the int() cast, it doesn't work in some cases when exporting to ONNX
+                positions = input.data.new(1, 1).fill_(int(self.padding_idx + input.size(1)))
+            else:
+                mask = input.data.ne(self.padding_idx).int()
+                positions = ( torch.cumsum(mask, dim=1).type_as(mask) * mask ).long() + self.padding_idx
+            real_positions = positions
+        else:
+            real_positions = positions
+        return super().forward(positions), real_positions
```

Comment:
It's a bit confusing to me that two tensors are returned -> could you add a comment explaining why this is the case? 
I think I would also prefer to split up the class into a function and a class:

1) Function that calculates the "real_positions" 
2) A very simple `nn.Embedding` 

I not really sure that we need such a big class. IMO, one can calculate the `real_positions` via a function, then we would only need `nn.Embedding` which is much more readable and don't need `_forward` at all and `max_positions` is such a small function that we can just copy paste it into the code. 
But maybe I overlooked something - What do you think? 

It's not that big of a deal though...we can also handle this in a refactor later.